devlib/instrument/monsoon.py

```diff
@@ -0,0 +1,170 @@
+
+	if report == 0:
+		popen_args = [MonsoonInstrument.MONSOON_BIN]
+		if self.device_entry:
+			popen_args += ['--device', self.device_entry]
+		popen_args += args
+
+	# Give up after 20 trials for !stream case (every attempt is timeout seconds)
+	n_trys_max = 20
+	n_trys = n_trys_max
+
+	while n_trys > 0:
+		# When reporting, don't try to run monsoon
+		if report == 0:
+			msg = "Running monsoon command: " + str(popen_args)
+			self.log_info(msg)
+			p = Popen(popen_args, stdout=PIPE, stderr=PIPE)
+
+		if async_mode == 1:
+			self.process = p
+			return "ASYNC"
+
+		if report == 1:
+			stream = 1
+			p = self.process
+			if p == None:
+				self.log_error("Monsoon not running") 
+			p.kill()
+		else:
+			# In sync mode (async_mode = 0), we kill monsoon after sometime
+			t = Timer(timeout, self.kill_monsoon, [p])
+			t.start()
+
+		# Get return code for task after either it quit or it was killed
+		waitret = p.wait()
+		if waitret != 0:
+			if report == 0:
+				t.cancel()
+
+			# If we received an error and we're !stream, do a retry
+			if stream == 0:
+				n_trys = n_trys - 1
+				errmsg="timed out waiting or error for monsoon, try resetting it"
+				self.log_error(errmsg)
+
+			errmsg = p.stderr.read()
+			stderrmsg = ""
+			for l in errmsg.split(os.linesep):
+				if l == "":
+					continue
+				stderrmsg += "STDERR: " + l + os.linesep
+
+			# Always print error buffer
+			self.log_error(stderrmsg)
+
+			if stream == 1:
+				# if stream, we don't consider -9 as an error, its just test completion timeout
+				# if we do return for any other reason, raise an exception to catch it later
+				if waitret != -9:
+					self.log_error(stderrmsg, exception=1)
+				self.log_info("MonsoonInstrument stream completed, returning")
+				return p.stdout.read()
+
+			if n_trys == 0:
+				errmsg = "MonsoonInstrument maximum retries exceeded, please POWER CYCLE your monsoon."
+				self.log_error(errmsg, exception=1)
+
+			errmsg = ("Retrying attempt " + str(n_trys_max - n_trys)) + ", last output:\n" + p.stdout.read() + "\n"
+			self.log_error(errmsg)
+		else:
+			if report == 0:
+				t.cancel()
+			self.log_info("MonsoonInstrument commands completed, returning")
+			return p.stdout.read()
+	self.log_error("monsoon unreachable code path", exception=1)
+
+    def get_status(self):
+        return self.run_monsoon(['--status'], timeout=3)
+
+    def set_voltage(self, volt):
+	if (volt > 5):
+		self.log_error("Too high voltage requested: " + str(volt))
+	return self.run_monsoon(['--voltage', str(volt)])
+
+    def set_current(self, current):
+	return self.run_monsoon(['--current', str(current)])
+
+    def set_start_current(self, current):
+	return self.run_monsoon(['--startcurrent', str(current)])
+
+    # Returns pandas dataframe for monsoon output (time and current in mA)
+    def get_samples_sync(self, duration):
+	txt = self.run_monsoon(['--hz', '5', '--timestamp', '--samples', '2000000'], stream=1, timeout=duration)
```

Comment:
These params should probably be exported somehow. Perhaps, since we have a single channel, we can specify them in the `__init__`?

In that case you can use them to configure the single InstrumentChannel which is going to be registered, as part of the instrument specific `attrs`:
https://github.com/ARM-software/devlib/blob/master/devlib/instrument/__init__.py#L142